packages/framework/ai-collab/src/explicit-strategy/idGenerator.ts

```diff
@@ -0,0 +1,90 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, oob } from "@fluidframework/core-utils/internal";
+import { Tree, NodeKind } from "@fluidframework/tree/internal";
+import type {
+	TreeNode,
+	ImplicitFieldSchema,
+	TreeArrayNode,
+	TreeFieldFromImplicitField,
+} from "@fluidframework/tree/internal";
+
+/**
+ * Given a tree node, generates a set of LLM friendly, unique ids for each node in a given Shared Tree.
+ * @remarks - simple id's are important for the LLM and this library to create and distinguish between different types certain TreeEdits
+ */
+export class IdGenerator {
+	private readonly idCountMap = new Map<string, number>();
+	private readonly prefixMap = new Map<string, string>();
+	private readonly nodeToIdMap = new Map<TreeNode, string>();
+	private readonly idToNodeMap = new Map<string, TreeNode>();
+
+	public constructor() {}
+
+	public getOrCreateId(node: TreeNode): string {
+		const existingID = this.nodeToIdMap.get(node);
+		if (existingID !== undefined) {
+			return existingID;
+		}
+
+		const schema = Tree.schema(node).identifier;
+		const id = this.generateID(schema);
+		this.nodeToIdMap.set(node, id);
+		this.idToNodeMap.set(id, node);
+
+		return id;
+	}
+
+	public getNode(id: string): TreeNode | undefined {
+		return this.idToNodeMap.get(id);
+	}
+
+	public getId(node: TreeNode): string | undefined {
+		return this.nodeToIdMap.get(node);
+	}
+
+	public assignIds(node: TreeFieldFromImplicitField<ImplicitFieldSchema>): string | undefined {
+		if (typeof node === "object" && node !== null) {
+			const schema = Tree.schema(node as unknown as TreeNode);
+			if (schema.kind === NodeKind.Array) {
+				(node as unknown as TreeArrayNode).forEach((element) => {
+					this.assignIds(element);
+				});
+			} else {
+				// TODO: SharedTree Team needs to either publish TreeNode as a class to use .instanceof() or a typeguard.
```

Comment:
@taylorsw04 are you guys tracking this?

Reply:
yup!

Reply:
Do you have a work item number we could reference in the TODO comment here?

Reply:
@taylorsw04